### PR TITLE
[invoke_subgraph] Fix get_output_metadata requires_grad bug

### PR DESCRIPTION
**Pull request by tugsbayasgalan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #178528

`meta["val"]` is always populated via `snapshot_fake → detach()`, which
strips `requires_grad`. This caused `get_output_metadata` to incorrectly
mark all float outputs as no-grad (0 tangents in backward) when taking
the static metadata path (e.g. in `reenter_make_fx`).

For float/complex tensor outputs, fall back to
`_get_output_metadata_by_execution` which checks `requires_grad` on the
actual executed output.

Authored with Claude.

### CODE REVIEW

File: torch/_higher_order_ops/invoke_subgraph.py [hunk 1]
```diff
@@ -469,8 +469,16 @@ def get_output_metadata(subgraph, *operands):
             output_metadata.indexes_with_symint.add(idx)
             output_metadata.indexes_with_no_grad.add(idx)
         elif isinstance(val, torch.Tensor):
-            # Check if tensor requires grad from metadata
-            if hasattr(val, "requires_grad") and not val.requires_grad:
+            if val.dtype.is_floating_point or val.is_complex():
+                # Can't trust requires_grad from meta["val"] since snapshot_fake
+                # uses detach() which strips it. Fall back to execution to get
+                # the accurate requires_grad for differentiable dtypes.
+                return _get_output_metadata_by_execution(subgraph, *operands)
```
Comment (ydwu4) [inline torch/_higher_order_ops/invoke_subgraph.py:476]: the time complexity of this can be too high?

Comment (tugsbayasgalan) [reply torch/_higher_order_ops/invoke_subgraph.py:476]: Hmmm this is already what happens in dynamo's path today i think.

Comment (ydwu4) [reply torch/_higher_order_ops/invoke_subgraph.py:476]: can they be merged into one execution outside of the loop?

File: test/test_proxy_tensor.py [hunk 2]
```diff
@@ -999,6 +999,26 @@ def _trace(f, *args):
     inps = [torch.randn(arg) for arg in args]
     return make_fx(f, tracing_mode="symbolic")(*inps)
 
+class TestGetOutputMetadata(TestCase):
+    def test_get_output_metadata_float_not_no_grad(self):
+        # Verify that get_output_metadata does not mark float tensor outputs
```
Comment (ydwu4) [inline test/test_proxy_tensor.py:1004]: remove the docs? the test is pretty clear by itself IMO

Comment (tugsbayasgalan) [reply test/test_proxy_tensor.py:1004]: yep
